### PR TITLE
Adds cache_data function for caching values on the master

### PR DESCRIPTION
**Pull request by ehelms**
Allows retrieving data from a cache file, or creates it with
supplied data if the file doesn't exist. This is useful for having
data that's randomly generated once on the master side (e.g. a password),
but then stays the same on subsequent runs.

### CODE REVIEW

**Comment by ehelms** [comment]
Another design option I contemplated was specifying the namespace in the name declaration and parsing it, e.g. cache_data('mysql/mysql_password', 'my_password')

**Comment by igalic** [comment]
that sounds interesting…

can you the errors for puppet future parser?

File: lib/puppet/parser/functions/cache_data.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    cache_dir_name = (args.size === 3) ? args[2] : 'data_cache'
+
+    cache_dir = File.join(Puppet[:vardir], cache_dir_name)
+    cache = File.join(cache_dir, name)
+
+    if File.exists? cache
+      YAML.load(File.read(cache))
+    else
+      FileUtils.mkdir_p cache_dir unless File.exists? cache_dir
```
Comment (ekohl) [inline lib/puppet/parser/functions/cache_data.rb]: Why the ```File.exists?``` I thought ```mkdir_p``` meant ignore the error if it already exists. Note I'm far from a ruby expert so maybe this is normal, but it feels weird to me.

File: lib/puppet/parser/functions/cache_data.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      File.open(cache, 'w', 0600) do |c|
+        c.write(YAML.dump(initial_data))
+      end
+      # Chown to puppet to prevent later cache-read errors when this is run as root
+      if Etc.getpwuid.name == 'root' && (uid = (Etc.getpwnam('puppet').uid rescue nil))
```
Comment (ekohl) [inline lib/puppet/parser/functions/cache_data.rb]: Maybe copy the uid from whatever ```Puppet[:vardir]``` has?

File: lib/puppet/parser/functions/cache_data.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+require 'yaml'
+require 'etc'
+
+# Retrieves data from a cache file, or creates it with supplied data if the file doesn't exist
+#
+# Useful for having data that's randomly generated once on the master side (e.g. a password), but
+# then stays the same on subsequent runs.
+#
+# Usage: cache_data(name, initial_data)
+# Example: $password = cache_data("mysql_password", 'my_password', 'mysql')
```
Comment (domcleal) [inline lib/puppet/parser/functions/cache_data.rb]: extra argument in this example than in the usage above?

**Comment by domcleal** [comment]
> Another design option I contemplated was specifying the namespace in the name declaration and parsing it, e.g. cache_data('mysql/mysql_password', 'my_password')

Or just make the namespace mandatory as the first argument?  I'd say the less optional behaviour, the better generally.  Either that or your suggestion will make the three-argument version more readable, as having the namespace as the last argument, *after* the key, reads strangely to me.

**Comment by ehelms** [comment]
Updated to:

1) Change to 3 argument design (and enforce namespace): namespace, data_name, initial_data
2) Use uid of the Puppet[:vardir]
3) Add mention of function to README

File: README.markdown [hunk 4]
```diff
@@ -118,6 +118,21 @@ echo($v7) # no comment here
 # My structure (Hash) {"b"=>["1", "2", "3"]}
 # (String) "12345"
 ```
+
+###cache_data
+
+Retrieves data from a cache file, or creates it with supplied data if the file doesn't exist. Useful for having data that's randomly generated once on the master side (e.g. a password), but then stays the same on subsequent runs.
+
```
Comment (nibalizer) [inline README.markdown]: Can you document what each parameter does here? This function is a bit magic so being explicit it a good idea I think.

**Comment by nibalizer** [comment]
I have a docs nitpick but once that's fixed and the tests working then I'm happy to merge this.

**Comment by igalic** [comment]
meanwhile travis is still breaking :(

**Comment by ehelms** [comment]
Tests passing and README updated

**Comment by ekohl** [comment]
One thing that may be tricky is that this has the same name as puppet-foreman's cache_data but different arguments so we need to think of a migration path.

**Comment by igalic** [comment]
gosh, yeah, that's something i didn't think of at all, even though i've discussed it so many times with @hlindberg

**Pull request merged**